### PR TITLE
LSNBLDR-790 checklist improvements (biggest is linking)

### PR DESCRIPTION
**Pull request by master-bob**
Big Jira:
https://jira.sakaiproject.org/browse/LSNBLDR-790

Left it in separate commits to show the individual changes made. (as well as to preserve author attribution)

### CODE REVIEW

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/util/SimplePageItemUtilities.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    /**
+     * Get's a name to display to the user from a SimplePageItem
+     * @param name, SimplePageItem
+     * @return String of the name
+     */
+    public static String getDisplayName(final SimplePageItem name) {
+        String itemName = name.getName();
+        if(itemName.isEmpty()) {
+            if(!name.getHtml().isEmpty()) {
+                itemName = Jsoup.parse(name.getHtml()).text();
```
Comment (buckett) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/util/SimplePageItemUtilities.java]: We already have `org.sakaiproject.util.api.FormattedText#stripHtmlFromText(java.lang.String, boolean)` in Sakai, is there a reason for using Jsoup instead?
I can image that JSoup is a better implementation, but consistency in a large project like Sakai can make things easier to understand and configure.

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java [hunk 2]
```diff
@@ -1291,19 +1314,19 @@ public void setAddChecklistItemData(String data) {
...
+			final Long index = (Long) checklistItem.get("index");
 
-		// We store with the index so that we can maintain the order
-		// in which the instructor inputted the checklist items
-		checklistItems.put(index, data);
+			// We store with the index so that we can maintain the order
+			// in which the instructor inputted the checklist items
+			checklistItems.put(index, data);
+
+		} catch (ParseException e) {
```
Comment (buckett) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java]: Could this also throw ClassCastException if the parser returns a JSONArray?

File: lessonbuilder/tool/pom.xml [hunk 3]
```diff
@@ -201,7 +201,11 @@
              <groupId>net.sf.opencsv</groupId>
              <artifactId>opencsv</artifactId>
         </dependency>
-
+        <dependency>
+            <groupId>org.jsoup</groupId>
+            <artifactId>jsoup</artifactId>
+            <version>LATEST</version>
```
Comment (buckett) [inline lessonbuilder/tool/pom.xml]: Is there a reason for this? As it leads to unreproducible builds where jsoup releases a new version and all of a sudden Sakai stops building.

Comment (master-bob) [inline lessonbuilder/tool/pom.xml]: O_o That wouldn't be good; I hadn't considered that. Will fix.

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/util/SimplePageItemUtilities.java (hunk 1 above)
Comment (master-bob) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/util/SimplePageItemUtilities.java]: I wasn't aware of that. I'll update to use that, and remove the Jsoup dependency.

File: lessonbuilder/tool/pom.xml (hunk 3 above)
Comment (master-bob) [inline lessonbuilder/tool/pom.xml]: That's a great point I hadn't considered!

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/util/SimplePageItemUtilities.java (hunk 1 above)
Comment (master-bob) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/util/SimplePageItemUtilities.java]: I'd have to use the static deprecated function: `org.sakaiproject.util.FormattedText#stripHtmlFromText(java.lang.String, boolean, boolean)` in order to remove the JSoup dependency.

JSoup is all ready being used in two places in Sakai. One in the [kernel-impl](https://github.com/sakaiproject/sakai/blob/476073dbf44548da46133cc61634cd3b1a5ea801/kernel/kernel-impl/pom.xml#L261) and one in [blackboard common import parser](https://github.com/sakaiproject/sakai/blob/8b1817a36a1da551527b41c09297ed3a7d7ca2b3/common/import-parsers/blackboard_6/impl/pom.xml#L36).

With some guidance, I would be happy to consolidate the use of JSoup across Sakai.

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java (hunk 2 above)
Comment (master-bob) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java]: Yes, that seems to be possible. It'll be caught and logged.

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/util/SimplePageItemUtilities.java (hunk 1 above)
Comment (buckett) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/util/SimplePageItemUtilities.java]: Yeah, the static `FormattedText` class should really not be a service as it means from a static utility method you can't easily call it.

I don't see the jSoup dependency as a blocker, but if you want to remove it you can do: `ComponentManager.get(FormattedText.class).stripHtmlFromText()`

**Comment by master-bob** [comment]
Forced pushed to resolve JSoup dependency.

**Comment by jonespm** [comment]
@master-bob This looks fine to us, but we really don't want to rebase all 8 commits as that could make future merging and tracking more work. 

If you wanted to retain the 2 authors could you rebase this locally to squash it down? Otherwise we can just squash/merge it to 1 commit if you aren't concerned about the individual authorship.

**Comment by master-bob** [comment]
I will squash this down to a commit per author. I left than separate so that the individual changes where easier to grok.

**Comment by master-bob** [comment]
@jonespm, I've squashed it down to two commits.

**Comment by master-bob** [comment]
Is there any hold up on this?

**Comment by jonespm** [comment]
No, I just missed the email from last week.

**Comment by master-bob** [comment]
It's quite all right. Thanks!

**Comment by jonespm** [comment]
And of course I rush and squash it :( :(

**Comment by master-bob** [comment]
I don't think @davidpbauer is too concerned with the attribution. And it's still in the PR itself, so in a sense the attribution is there to some degree.

**Comment by jonespm** [comment]
Yeah sorry @davidpbauer !

**Pull request merged**